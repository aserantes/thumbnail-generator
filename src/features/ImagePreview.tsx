import React from "react";
import {
  ButtonRow,
  ImageInfo,
  ImageInfoTextBlock,
  ImageInfoTextItem,
  ImagePreviewWrapper,
  Thumbnail,
} from "../components/Common";
import { ProcessButton, RestartButton } from "../components";
import { getName, getPath, getSize, getType } from "../store/selectors";
import { useSelector } from "react-redux";

export function ImagePreview() {
  const name = useSelector(getName);
  const type = useSelector(getType);
  const size = useSelector(getSize);
  const path = useSelector(getPath);
  const sizeText = `${(size / 1024).toFixed(2)} Kb`;

  return (
    <ImagePreviewWrapper data-testid="ImagePreview-wrapper">
      <ImageInfo>
        <Thumbnail alt="Image Preview" src={path} />
        <ImageInfoTextBlock>
          <ImageInfoTextItem>
            <strong>Name:&nbsp;</strong> {name}
          </ImageInfoTextItem>
          <ImageInfoTextItem>
            <strong>Type:&nbsp;</strong> {type}
          </ImageInfoTextItem>
          <ImageInfoTextItem>
            <strong>Size:&nbsp;</strong> {sizeText}
          </ImageInfoTextItem>
        </ImageInfoTextBlock>
      </ImageInfo>
      <ButtonRow>
        <RestartButton />
        <ProcessButton />
      </ButtonRow>
    </ImagePreviewWrapper>
  );
}
/*
const onSubmit = (e) => {
  e.preventDefault();
  const formData = new FormData();
  formData.append("file", file);

  // Pass a setUploadPercentage as callback here
  putData(formData, { onUploadProgress: setUploadPercentage )
    .then((response) => { 
      console.log(response);
    });
};
export function putData(formData, { onUploadProgress }) {
  return axios.put('/update', formData, {
    headers: {
      "Content-Type": "multipart/form-data",
    },
    onUploadProgress : (progressEvent) => {
      const progress = parseInt(Math.round((progressEvent.loaded * 100) / progressEvent.total));
      // Update state here
      onUploadProgress(progress);
    },
  });
}

*/
