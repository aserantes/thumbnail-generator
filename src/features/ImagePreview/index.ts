export { ImagePreview } from "./ImagePreview";

export {
  initialState,
  imagePreviewReducer,
  changeTest,
} from "./ImagePreviewSlice";
