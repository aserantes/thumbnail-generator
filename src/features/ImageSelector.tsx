import React, { useEffect } from "react";
import {
  BrowseButton,
  ButtonRow,
  CameraButton,
  DropZone,
  ImageSelectorWrapper,
} from "components";
import { getFileInfo } from "utils/fileHelpers";
import { selectors } from "store";
import { setFileToUploadData } from "store/fileToUploadSlice";
import { useDispatch, useSelector } from "react-redux";
import { useFileTypeAnalizer } from "hooks/useFileTypeAnalizer";

export function ImageSelector() {
  const chunkPath = useSelector(selectors.getChunkPath);
  const [result] = useFileTypeAnalizer(chunkPath);
  const dispatch = useDispatch();
  const handleDrop = (event: React.DragEvent<HTMLDivElement>): void => {
    const file = event.dataTransfer.files && event.dataTransfer.files[0];
    const fileInfo = getFileInfo(file);

    fileInfo && dispatch(setFileToUploadData(fileInfo));
  };
  const handleChange = (event: React.ChangeEvent<HTMLInputElement>): void => {
    const file = event.target.files && event.target.files[0];
    const fileInfo = getFileInfo(file);

    fileInfo && dispatch(setFileToUploadData(fileInfo));
  };

  useEffect(() => {
    if (result.ext !== "") {
      dispatch(
        setFileToUploadData({
          type: result.ext,
        })
      );
    }
  }, [result, dispatch]);

  return (
    <ImageSelectorWrapper data-testid="ImageSelector-wrapper">
      <DropZone onFileDrop={handleDrop} />
      <ButtonRow>
        <BrowseButton onFileChange={handleChange} />
        <CameraButton onFileChange={handleChange} />
      </ButtonRow>
    </ImageSelectorWrapper>
  );
}
