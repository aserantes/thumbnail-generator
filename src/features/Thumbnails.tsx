import React from "react";
import thumb120 from "../thumbnails/thumb-120x120.png";
import thumb160 from "../thumbnails/thumb-160x120.png";
import thumb400 from "../thumbnails/thumb-400x300.png";
import { ImageInfo, Thumbnail2, ThumbnailsWrapper } from "../components/Common";

export function Thumbnails() {
  return (
    <ThumbnailsWrapper data-testid="Thumbnails-wrapper">
      <ImageInfo>
        <a download href={thumb120}>
          <Thumbnail2
            alt="Image Preview"
            height="120px"
            src={`/thumbnails/thumb-120x120.png?t=${Date.now()}"`}
            width="120px"
          />
        </a>
      </ImageInfo>
      <ImageInfo>
        <a download href={thumb160}>
          <Thumbnail2
            alt="Image Preview"
            height="120px"
            src={`/thumbnails/thumb-160x120.png?t=${Date.now()}"`}
            width="160px"
          />
        </a>
      </ImageInfo>
      <ImageInfo>
        <a download href={thumb400}>
          <Thumbnail2
            alt="Image Preview"
            height="300px"
            src={`/thumbnails/thumb-400x300.png?t=${Date.now()}"`}
            width="400px"
          />
        </a>
      </ImageInfo>
    </ThumbnailsWrapper>
  );
}
