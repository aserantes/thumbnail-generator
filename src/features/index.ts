export { ImagePreview } from "./ImagePreview";

export { ImageSelector } from "./ImageSelector";

export { Thumbnails } from "./Thumbnails";
