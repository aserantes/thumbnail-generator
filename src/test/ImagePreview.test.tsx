import React from "react";
import { ImagePreview } from "../features";
import { renderWithRedux, screen } from "../helpers";

describe("<ImagePreview />", () => {
  it("should render", () => {
    renderWithRedux(<ImagePreview />);

    const component = screen.getByTestId("ImagePreview-wrapper");

    expect(component).toBeInTheDocument();
  });

  it("should match snapshot", () => {
    expect(renderWithRedux(<ImagePreview />).asFragment()).toMatchSnapshot();
  });

  it("should show the image preview", () => {});

  it("should show the filename", () => {});

  it("should show the filesize", () => {});

  it("should show the filetype", () => {});

  it("should show a button with text 'Process'", () => {});

  it("should call 'processFile' with file info on 'Process' button click", () => {}); // In my case, 'processFile' will be a redux action creator which will dispatch an async thunked redux action, but I want this TDD approach test cases description to be as agnostic of implementation as possible

  it("should show a disabled button with text 'Processing...' after calling 'processFile'", () => {});

  it("should show error on status rejected", () => {});
});

/*  
  it("should match snapshot and div contain reduxStringInitialState", () => {
    const { asFragment } = renderWithRedux(<ImagePreview />);
    const fragment = asFragment();

    expect(fragment).toMatchSnapshot();

    expect(screen.getByText("reduxStringInitialState")).toBeInTheDocument();
  });

  it("REDUX:should call setFileToUploadData action on click //// UI:and change div value", () => {
    const { mockStore } = renderWithRedux(<ImagePreview />);

    fireEvent.click(screen.getByText("change test"));

    expect(mockStore.getActions()).toEqual([
      { type: setFileToUploadData.type, payload: "" },
    ]);

    // SEPARAR REDUX ACTION REDUCER TESTING FROM UI USER BEHAVIOUR
    expect(screen.getByText("reduxStringInitialState")).not.toBeInTheDocument();
  });

  it("should call setFileToUploadData on click", () => {
    const { mockStore } = renderWithRedux(<ImagePreview />);

    const input = screen.getByTestId("input");

    fireEvent.change(input, { target: { value: "123" } });

    fireEvent.click(screen.getByTestId("button"));

    expect(mockStore.getActions()).toEqual([
      { type: setFileToUploadData.type, payload: "123" },
    ]);
  });

  test("fileToUploadReducer should change test value", () => {
    expect(
      fileToUploadReducer(
        { ...rootInitialState.imagePreview, test: "" },
        setFileToUploadData("hola")
      )
    ).toEqual({ ...rootInitialState.imagePreview, test: "hola" });
  });

*/
