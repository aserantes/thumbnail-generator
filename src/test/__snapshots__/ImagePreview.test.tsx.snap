// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ImagePreview /> should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="sc-crzoUp sc-fnVYJo sc-bkbjAj eoKvYH gJOtFr fCCFxd"
    data-testid="ImagePreview-wrapper"
  >
    <div
      class="sc-hKFyIo ihyQnA"
    >
      <img
        alt="Image Preview"
        class="sc-gKAblj jWASIP"
        src=""
      />
      <div
        class="sc-eCApGN eDeRTi"
      >
        <div
          class="sc-jSFkmK hUSRvy"
        >
          <strong>
            Name: 
          </strong>
           
        </div>
        <div
          class="sc-jSFkmK hUSRvy"
        >
          <strong>
            Type: 
          </strong>
           
        </div>
        <div
          class="sc-jSFkmK hUSRvy"
        >
          <strong>
            Size: 
          </strong>
           0.00 Kb
        </div>
      </div>
    </div>
    <p
      class="sc-jrsJCI sc-kEqYlL fTYhRe bPdPvr"
      hidden=""
    />
    <div
      class="sc-hBMVcZ iQQiiH"
    >
      <div
        class="sc-bdnylx elazSQ"
        role="button"
      >
        <div
          class="sc-gtssRu jFSpxo"
        >
          <i
            class="fas fa-step-backward"
          />
        </div>
        <div
          class="sc-pNWxx cpZoGq"
        >
          Restart
        </div>
      </div>
      <div
        class="sc-bdnylx elazSQ"
        role="button"
      >
        <div
          class="sc-gtssRu jFSpxo"
        >
          <i
            class="fas fa-thumbs-up"
          />
        </div>
        <div
          class="sc-pNWxx cpZoGq"
        >
          Process
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
